keps/sig-api-machinery/1886-jsonpath/README.md

```diff
@@ -0,0 +1,331 @@
+  - [util/jsonpath](#utiljsonpath)
+  - [forked/golang/template](#forkedgolangtemplate)
+  - [Server-side usage](#server-side-usage)
+  - [Client-side usage](#client-side-usage)
+  - [Test Plan](#test-plan)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [Graduation Criteria](#graduation-criteria)
+    - [Beta](#beta)
+    - [GA](#ga)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [References](#references)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+<!--
+**ACTION REQUIRED:** In order to merge code into a release, there must be an
+issue in [kubernetes/enhancements] referencing this KEP and targeting a release
+milestone **before the [Enhancement Freeze](https://git.k8s.io/sig-release/releases)
+of the targeted release**.
+
+For enhancements that make changes to code or processes/procedures in core
+Kubernetes i.e., [kubernetes/kubernetes], we require the following Release
+Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+- [ ] Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+- [ ] KEP approvers have approved the KEP status as `implementable`
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+[original JSONPath syntax]: https://goessner.net/articles/JsonPath
+[existing JSONPath doc]: https://github.com/kubernetes/website/blob/ecc27bbbe70f92d031fa52be76bb9471b2e83152/content/en/docs/reference/kubectl/jsonpath.md
+[JSON Strategic Merge Patch]: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-api-machinery/strategic-merge-patch.md
+[go template]: https://golang.org/pkg/text/template/
+## Summary
+
+This proposal aims to clarify the JSONPath template syntax supported by Kubernetes
+and make some enhancement to meet today's user needs.
+
+In addition to [original JSONPath syntax], this JSONPath enhancement mainly allow users to apply more complex, flexible filter
```

Comment:
Take a look at JSONPointer's spec: https://tools.ietf.org/html/rfc6901

I really appreciate that you're defining a spec here, I think this is the right thing to do for JSONPath. However I'm not sure that this is the best forum for this, and as much as I'd love to nerd out about it, I don't really have time :( I wonder if it would be better to attempt to work out the JSONPath spec in a standards body?

Reply:
Thanks for your timely review.

> Take a look at JSONPointer's spec: https://tools.ietf.org/html/rfc6901

I made this spec by referring KEP template and several submitted KEPs.
I wonder if you are worrying about format of the spec?

> I'm not sure that this is the best forum for this. I wonder if it would be better to attempt to work out the JSONPath spec in a standards body?

Haha. I am sure this is the "correct" forum for making JSONPath enhancement for Kubernetes, though it might not be a ideal place for JSONPath spec itself.
I once considered pushing a RFC standard for JSONPath, but the road is too long to see the end, makes me feel exhausted.
Since we are discussing a super-set of JSONPath, I think we can treat it as a enhancement-proposal.
Making spec here is more about clarifying syntax supported by Kubernetes and tracking changes in application lifecycle.

I personally prefer the Kubernetes-style rather than RFC-style. It's more clear, applicable and human-readable as a reference while using and developing Kubernetes.